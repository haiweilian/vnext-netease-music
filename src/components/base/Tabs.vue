<template>
  <ul class="tabs">
    <li
      v-for="tab of tabs"
      :key="tab.value"
      :class="['tabs__item', tab.value === currentValue ? 'is-active' : '']"
      @click="setCurrentValue(tab.value)"
    >
      <span>{{ tab.label }}</span>
    </li>
  </ul>
</template>

<script setup lang="ts">
import { ref, toRef, watch } from 'vue'
import type { PropType } from 'vue'
import type { ISingleTab } from '~/types'

const props = defineProps({
  tabs: {
    type: Object as PropType<ISingleTab[]>,
    required: true,
  },
  modelValue: {
    type: [String, Number],
    default: '',
  },
})
const tabs = toRef(props, 'tabs')
const emits = defineEmits(['update:modelValue', 'change'])

/**
 * 切换通知更新值和事件
 */
const currentValue = ref<string | number>(props.modelValue)
const setCurrentValue = (value: string | number) => {
  currentValue.value = value
  emits('update:modelValue', currentValue.value)
  emits('change', currentValue.value)
}

/**
 * 监听绑定值赋值默认值
 */
watch(() => props.modelValue, (modelValue) => {
  currentValue.value = modelValue
})
</script>

<style lang="scss" scoped>
@include b(tabs) {
  display: flex;
  justify-content: flex-end;
  list-style: none;

  @include e(item) {
    padding: 12px 0;
    margin: 0 12px;
    font-size: 12px;
    color: #7f7f81;
    white-space: nowrap;
    cursor: pointer;

    @include when(active) {
      color: #d33a31;
    }

    &:hover {
      color: #000;
    }
  }
}
</style>
