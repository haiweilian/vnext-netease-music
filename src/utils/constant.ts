/**
 * 全局用户标识名称
 */
export const GLOBAL_UID_KEY = 'vnext-netease-music-uid'

/**
 * 全局搜索历史
 */
export const GLOBAL_SEARCH_HOT_KEY = 'vnext-netease-music-search-hot'

/**
 * 订阅/取消订阅
 * 1: 订阅、0: 取消
 */
export enum SubAction {
  sub = '1',
  unsub = '0',
}

/**
 * 评论类型
 * 0: 歌曲、1: mv、2: 歌单、3: 专辑、4: 电台、5: 视频、6: 动态
 */
export enum CommentType {
  song = '0',
  mv = '1',
  playlist = '2',
  album = '3',
  dj = '4',
  video = '5',
  event = '6',
}

/**
 * 搜索类型
 * 1: 单曲、 10: 专辑、 100: 歌手、 1000: 歌单、 1002: 用户、 1004: MV、 1006: 歌词、 1009: 电台、 1014: 视频
 */
export enum SearchType {
  single = '1',
  album = '10',
  artist = '100',
  playlist = '1000',
  user = '1002',
  mv = '1004',
  lyric = '1006',
  dj = '1009',
  video = '1014',
  complex = '1018',
}
