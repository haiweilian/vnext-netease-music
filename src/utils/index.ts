/**
 * 是否为空
 */
export function isEmpty(value: any) {
  return value === undefined || value === null || value === ''
}

/**
 * 控制图片尺寸
 * @example http://p4.music.126.net/JzNK4a5PjjPIXAgVlqEc5Q==/109951164154280311.jpg?param=200y200
 */
export function thumbnail(url: string, w: number, h: number = w) {
  if (url) {
    return `${url}?param=${w}y${h}`
  }
  else {
    return `https://p2.music.126.net/6y-UleORITEDbvrOLV0Q8A==/5639395138885805.jpg?param=${w}y${h}`
  }
}

/**
 * 不足十补零
 */
export function padZero(value: number) {
  return value < 10 ? `0${value}` : `${value}`
}

/**
 * 数量转换单位
 */
export function formatCount(value: number) {
  value = Number(value) || 0
  return value > 100000 ? `${Math.round(value / 10000)} 万` : value
}
