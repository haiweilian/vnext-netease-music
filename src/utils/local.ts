import type { IMenu, ISingleTab } from '~/types'

/**
 * 左侧菜单栏
 */
export const localMenus: IMenu[] = [{
  name: '',
  children: [
    { name: '发现音乐', icon: 'music', link: '/discovery' },
    { name: '推荐歌单', icon: 'playlist', link: '/playlist' },
    { name: '最新音乐', icon: 'song', link: '/song' },
  ],
}]

/**
 * 歌单分类
 */
export const localPlaylistTabs: ISingleTab[] = [
  { value: '全部', label: '全部' },
  { value: '欧美', label: '欧美' },
  { value: '华语', label: '华语' },
  { value: '流行', label: '流行' },
  { value: '说唱', label: '说唱' },
  { value: '摇滚', label: '摇滚' },
  { value: '民谣', label: '民谣' },
  { value: '电子', label: '电子' },
  { value: '轻音乐', label: '轻音乐' },
  { value: 'ACG', label: 'ACG' },
  { value: '怀旧', label: '怀旧' },
  { value: '治愈', label: '治愈' },
  { value: '旅行', label: '旅行' },
]

/**
 * 歌曲分类
 */
export const localSongTabs: ISingleTab[] = [
  { label: '全部', value: 0 },
  { label: '华语', value: 7 },
  { label: '欧美', value: 96 },
  { label: '日本', value: 8 },
  { label: '韩国', value: 16 },
]
