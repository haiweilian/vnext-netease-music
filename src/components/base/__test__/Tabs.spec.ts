import { nextTick } from 'vue'
import { mount } from '@vue/test-utils'
import Tabs from '../Tabs.vue'
import { localPlaylistTabs } from '~/utils/local'

describe('Tabs', () => {
  test('should tabs props', async () => {
    // first
    const first = localPlaylistTabs[0].value
    const wrapper = mount(Tabs, {
      props: {
        tabs: localPlaylistTabs,
        modelValue: first,
      },
    })
    await nextTick()
    const firstElement = wrapper.findAll('li')[0]

    expect(firstElement.exists()).toBeTruthy()
    expect(firstElement.classes('is-active')).toBeTruthy()
    expect(firstElement.html()).toContain(first)
    expect(wrapper.html()).toMatchSnapshot()

    // change last
    const last = localPlaylistTabs[localPlaylistTabs.length - 1].value
    wrapper.setProps({
      modelValue: last,
    })
    await nextTick()
    const lastElement = wrapper.findAll('li')[localPlaylistTabs.length - 1]

    expect(lastElement.exists()).toBeTruthy()
    expect(lastElement.classes('is-active')).toBeTruthy()
    expect(lastElement.html()).toContain(last)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should tabs emits', async () => {
    const wrapper = mount(Tabs, {
      props: {
        tabs: localPlaylistTabs,
      },
    })

    wrapper.findAll('li')[0].trigger('click')
    await nextTick()

    // change event
    const changeEvent = wrapper.emitted('change') as any[]
    expect(changeEvent).toHaveLength(1)
    expect(changeEvent[0]).toEqual([localPlaylistTabs[0].value])

    // update event
    const updateEvent = wrapper.emitted('update:modelValue') as any[]
    expect(updateEvent).toHaveLength(1)
    expect(updateEvent[0]).toEqual([localPlaylistTabs[0].value])

    expect(wrapper.html()).toMatchSnapshot()
  })
})
