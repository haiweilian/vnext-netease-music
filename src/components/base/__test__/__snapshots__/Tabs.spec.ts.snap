// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should tabs emits 1`] = `
"<ul class=\\"tabs\\">
  <li class=\\"tabs__item is-active\\"><span>全部</span></li>
  <li class=\\"tabs__item\\"><span>欧美</span></li>
  <li class=\\"tabs__item\\"><span>华语</span></li>
  <li class=\\"tabs__item\\"><span>流行</span></li>
  <li class=\\"tabs__item\\"><span>说唱</span></li>
  <li class=\\"tabs__item\\"><span>摇滚</span></li>
  <li class=\\"tabs__item\\"><span>民谣</span></li>
  <li class=\\"tabs__item\\"><span>电子</span></li>
  <li class=\\"tabs__item\\"><span>轻音乐</span></li>
  <li class=\\"tabs__item\\"><span>ACG</span></li>
  <li class=\\"tabs__item\\"><span>怀旧</span></li>
  <li class=\\"tabs__item\\"><span>治愈</span></li>
  <li class=\\"tabs__item\\"><span>旅行</span></li>
</ul>"
`;

exports[`Tabs should tabs props 1`] = `
"<ul class=\\"tabs\\">
  <li class=\\"tabs__item is-active\\"><span>全部</span></li>
  <li class=\\"tabs__item\\"><span>欧美</span></li>
  <li class=\\"tabs__item\\"><span>华语</span></li>
  <li class=\\"tabs__item\\"><span>流行</span></li>
  <li class=\\"tabs__item\\"><span>说唱</span></li>
  <li class=\\"tabs__item\\"><span>摇滚</span></li>
  <li class=\\"tabs__item\\"><span>民谣</span></li>
  <li class=\\"tabs__item\\"><span>电子</span></li>
  <li class=\\"tabs__item\\"><span>轻音乐</span></li>
  <li class=\\"tabs__item\\"><span>ACG</span></li>
  <li class=\\"tabs__item\\"><span>怀旧</span></li>
  <li class=\\"tabs__item\\"><span>治愈</span></li>
  <li class=\\"tabs__item\\"><span>旅行</span></li>
</ul>"
`;

exports[`Tabs should tabs props 2`] = `
"<ul class=\\"tabs\\">
  <li class=\\"tabs__item\\"><span>全部</span></li>
  <li class=\\"tabs__item\\"><span>欧美</span></li>
  <li class=\\"tabs__item\\"><span>华语</span></li>
  <li class=\\"tabs__item\\"><span>流行</span></li>
  <li class=\\"tabs__item\\"><span>说唱</span></li>
  <li class=\\"tabs__item\\"><span>摇滚</span></li>
  <li class=\\"tabs__item\\"><span>民谣</span></li>
  <li class=\\"tabs__item\\"><span>电子</span></li>
  <li class=\\"tabs__item\\"><span>轻音乐</span></li>
  <li class=\\"tabs__item\\"><span>ACG</span></li>
  <li class=\\"tabs__item\\"><span>怀旧</span></li>
  <li class=\\"tabs__item\\"><span>治愈</span></li>
  <li class=\\"tabs__item is-active\\"><span>旅行</span></li>
</ul>"
`;
